src/hazelcore/OcamlExtraction.re

```diff
@@ -0,0 +1,374 @@
+    | Bool => "bool"
+    // add parenthesis is necessary
+    // (int -> int) -> int != int -> int -> int
+    | Arrow(t1, t2) =>
+      "(" ++ extract(~t=t1) ++ " -> " ++ extract(~t=t2) ++ ")"
+    | Sum(t1, t2) => "(" ++ extract(~t=t1) ++ " | " ++ extract(~t=t2) ++ ")"
+    | Prod(tl) => "(" ++ list_prod(~tl) ++ ")"
+    | List(t) => extract(~t) ++ " list"
+    }
+  and list_prod = (~tl: list(HTyp.t)): ocaml_typ =>
+    switch (tl) {
+    | [] => ""
+    | [h, ...t] => extract(~t=h) ++ " * " ++ list_prod(~tl=t)
+    };
+};
+
+module Pat = {
+  type ocaml_pat = string;
+  // should write like "expand"
+  let rec extract = (~dp: DHPat.t): ocaml_pat => {
+    switch (dp) {
+    | EmptyHole(_, _) => failwith("Pat: Empty Hole")
+    | NonEmptyHole(_, _, _, _) => failwith("Pat: NonEmptyHole")
+    | Wild => "_"
+    | Keyword(_, _, _) => failwith("Pat: Incomplete Program, Keyword")
+    | Var(s) => s
+    | IntLit(i) => string_of_int(i)
+    | FloatLit(f) => string_of_float(f)
+    | BoolLit(b) => string_of_bool(b)
+    | Inj(_side, t) => extract(~dp=t)
+    | ListNil => "[]"
+    | Cons(dp1, dp2) => extract(~dp=dp1) ++ "::" ++ extract(~dp=dp2)
+    | Pair(dp1, dp2) =>
+      "(" ++ extract(~dp=dp1) ++ ", " ++ extract(~dp=dp2) ++ ")"
+    | Triv => "()"
+    | Ap(dp1, dp2) => extract(~dp=dp1) ++ " " ++ extract(~dp=dp2)
+    };
+  };
+};
+
+module Exp = {
+  type ocaml_exp = string;
+
+  // (extracted expression, dynamic/inner type) like ExpandResult
+  // Htyp.t is like the ground type
+
+  //type extract_result = option((ocaml_exp, HTyp.t));
+  type extract_result = (ocaml_exp, HTyp.t);
+
+  /* Theorem: If extract(ctx, d) = Some(ce, ty) and ctx |- d : ty then ce :_ocaml Typ.extract(ty) */
+
+  //TODO: replace == with HTyp.consistent
+  let rec extract = (~ctx: Contexts.t, ~de: DHExp.t): extract_result =>
+    switch (de) {
+    | EmptyHole(_) => failwith("Exp: Empty Hole")
+    | NonEmptyHole(_) => failwith("Exp: Non-Empty Hole")
+    // used for variables that are keywords in Hazel, like "let" and "case"
+    | Keyword(_) => failwith("Exp: Incomplete Program (Keyword)")
+    | FreeVar(_, _, map, x) =>
+      //lookup the expression represented by x
+      let map_e = VarMap.lookup(map, x);
+      switch (map_e) {
+      | None => failwith("Exp: FreeVar Not Found")
+      | Some(e) =>
+        let ocaml_e = extract(~ctx, ~de=e);
+        //add a parenthesize to avoid any possible order priority
+        (
+          "(" ++ x ++ " : " ++ Typ.extract(~t=snd(ocaml_e)) ++ ")",
+          snd(ocaml_e),
+        );
+      };
+    | BoundVar(x) =>
+      //directly lookup type in environment
+      let typ = VarCtx.lookup(Contexts.gamma(ctx), x);
+      switch (typ) {
+      | None => failwith("Exp: BoundVar " ++ x ++ " Not Found")
+      // if the hole type, we don't print the type
+      | Some(t) =>
+        if (t == Hole) {
+          ("(" ++ x ++ ")", t);
+        } else {
+          ("(" ++ x ++ " : " ++ Typ.extract(~t) ++ ")", t);
```

Comment:
do you really need to annotate each variable?

Reply:
I also don't like that, because it will make the extracted code very messy... May I change to that we only annotate variables when they're declared?

Reply:
yes if you annotate every variable binding then the variable will necessarily have that type everywhere